src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php

```diff
@@ -29,13 +29,17 @@ class ProductRepository extends BaseProductRepository implements ProductReposito
      */
     public function createByTaxonPaginator(TaxonInterface $taxon, array $criteria = [])
     {
+        $root = $taxon->isRoot() ? $taxon : $taxon->getRoot();
+
         $queryBuilder = $this->createQueryBuilder('o');
         $queryBuilder
             ->innerJoin('o.taxons', 'taxon')
+            ->andWhere($queryBuilder->expr()->eq('o.root', ':root'))
```

Comment:
Anyway, for nested set tree, left / right conditions should be enough -  https://en.wikipedia.org/wiki/Nested_set_model. Moreover, we can replace this:
```
->andWhere($queryBuilder->expr()->orX(
    'taxon = :taxon',
    ':left < taxon.left AND taxon.right < :right'
))
```

with:

```
->andWhere(':left <= taxon.left AND taxon.right <= :right')
```

But there should be some automatic tests made before (if we need this method at all).